### PR TITLE
Assignment 10 august 22

### PR DESCRIPTION
**Pull request by HassamAzam**
https://github.com/user-attachments/assets/a057e32f-789b-443f-84ad-f4a5297d09c2

### CODE REVIEW

File: Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { Button } from "@mui/material";
+
+const ButtonComponent = ({
+  label,
+  onClick,
+  color = "primary",
+  variant = "contained",
+  ...props
+}) => {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js]: if we are simply returning the component then we can also use shorthand operator

```suggestion
}) => (
    <Button onClick={onClick} color={color} variant={variant} {...props}>
      {label}
    </Button>
  );
```

File: Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  return (
+    <Button onClick={onClick} color={color} variant={variant} {...props}>
+      {label}
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js]: what if we doon't pass any label, we should have add default parameters.

File: Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    </Button>
+  );
+};
+
+export default ButtonComponent;
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js:31]: for components like these, we should use `prop-types`.

File: Assignment-10-August-22/userprofile/src/Components/InputField.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import { TextField } from "@mui/material";
+const InputField = ({ label, placeholder, type = "text", ...props }) => {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/InputField.js]: add a line before that.

File: Assignment-10-August-22/userprofile/src/Components/InputField.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  return (
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/InputField.js]: same shorthand and prop-types

File: Assignment-10-August-22/userprofile/src/Store/authSlice.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+const initialState = {
+  userEmail: null,
+};
+const authSlice = createSlice({
+  name: "auth",
+  initialState,
+  reducers: {
+    login: (state, action) => {
+      state.userEmail = action.payload;
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Store/authSlice.js:46]: we shouldn't handle authentications in store, becuase when on page reload, the store get updated. use localstorage, sessionstorage

File: Assignment-10-August-22/userprofile/src/Store/userArraySlice.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  initialState: {
+    value: [],
+  },
+  reducers: {
+    addInArray: (state, action) => {
+      state.value.push(action.payload);
+    },
+    updateInArray: (state, action) => {
+      const { email, updatedData } = action.payload;
+      const index = state.value.findIndex((user) => user.email === email);
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Store/userArraySlice.js:35]: what if i updated the email field?

File: Assignment-10-August-22/userprofile/src/containers/App/App.css [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  background-color: white;
+  padding: 20px;
+  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
+  border-radius: 8px;
+  width: 300px; 
+}
+.red-toast-message{
+  background-color: red;
+  color: white;
+}
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/App/App.css]: EOF (end of file)

File: Assignment-10-August-22/userprofile/src/containers/App/App.css [hunk 9]
```diff
@@ -0,0 +1,40 @@
+.App {
+  text-align: center;
+}
+.formContainer {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/App/App.css:5]: add an extra line before that selector and do it for the rest of the file as well.

File: Assignment-10-August-22/userprofile/src/containers/App/App.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+function App() {
+  return (
+
+    <div className="App">
+    <Provider store={store}>
+        <Routes>
+          <Route path="/" element={<Login />} />
+          <Route path="/login" element={<Login />} />
+          <Route path="/signup" element={<SignUp />} />
+          <Route path="/dashboard" element={<Dashboard />} />
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/App/App.js]: Make another custom component, called `<PrivateRoute component={Dashboard} />` if any component present in privateComponent then it should pass / fullfil the authentication layer and then land on Private component.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+import { updateInArray } from "../../Store/userArraySlice";
+
+import ButtonComponent from "../../Components/ButtonComponent";
+import InputField from "../../Components/InputField";
+
+export default function Dashboard() {
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+
+  const userEmail = useSelector((state) => state.auth.userEmail);
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: try `reselect` library

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  const userArray = useSelector((state) => state.userArray.value);
+
+  const loggedInUser = userArray.find((user) => user.email === userEmail);
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: this data should direct come from selector and You have to comment that reason why?

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+  const validationSchema = Yup.object().shape({
+    firstName: Yup.string().required("First Name is required"),
+    lastName: Yup.string().required("Last Name is required"),
+    country: Yup.string().required("Country is required"),
+  });
+
+  const handleLogout = () => {
+    dispatch(logout());
+    navigate("/login");
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js:47]: that would work for now.
what if in future, we have to logout by destorying the session or we need to call a logout API, in that scenario, we will nagivate `/login` page and then come back to dasboard page, untill logout API response's return. 

Make these two statments sync with each other

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+          },
+        })
+      );
+      setEditing(false);
+    }
+  };
+
+  if (!userEmail) {
+    navigate("/login");
+    return null;
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: why return null?

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+import { useDispatch, useSelector } from "react-redux";
+import { toast, ToastContainer } from "react-toastify";
+import "react-toastify/dist/ReactToastify.css";
+import { useNavigate, Link } from "react-router-dom";
+import { FormGroup, FormControl } from "@mui/material";
+import { useForm, Controller } from "react-hook-form";
+import { yupResolver } from "@hookform/resolvers/yup";
+import * as Yup from "yup";
+
+import { login } from "../../Store/authSlice";
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: we only use Pascal cases for components, not store.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  const {handleSubmit,control,  formState: { errors }} = useForm({  
+    resolver: yupResolver(validationSchema),
+    defaultValues: {
+      email: "",
+      password: "",
+    },
+  });
+
+  const handleLogin = (values) => {
+    const user = userArray.find((user) => user.email === values.email);
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: should come from selector

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+    if (user) {
+      if (user.password === values.password) {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: use `?` notation for check weather the object is null or undefined. e.g

```suggestion
      if (user?.password === values?.password) {
```

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+        dispatch(login(user.email));
+        toast.success("Login Successful");
+
+        setTimeout(() => {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: why making UI block for 2 sec? what if i continously start pressing login button again and again?

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+const validationSchema = Yup.object({
+  email: Yup.string().email("Invalid email format").required("Email is required"),
+  password: Yup.string().min(6, "Password must be at least 6 characters").required("Password is required"),
+});
+
+export default function Login() {
+  const userArray = useSelector((state) => state.userArray.value);
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js:31]: I don't think you are setting document title anywhere. Make a custom function to handle that.

File: Assignment-10-August-22/userprofile/src/containers/Pages/SignUp.js [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+      password: "",
+      country: "",
+    },
+  });
+
+  const handleClose = () => {
+    navigate("/login");
+  };
+
+  const checkIfUserExists = (email) => {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/SignUp.js]: use shorthand return statement

File: Assignment-10-August-22/userprofile/src/containers/Pages/SignUp.js [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+    } else {
+      dispatch(addInArray(values));
+      navigate("/login");
+    }
+  };
+
+  return (
+    <div>
+      <Modal
+        open={true}
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/SignUp.js]: In React, if we don't need to explicity pass `true` as props. we simply use `open` without any props.

```
 <Modal
        open
        onClose={handleClose}
        aria-labelledby="registration-form"
        aria-describedby="registration-form-description"
/>
```

File: Assignment-10-August-22/userprofile/src/containers/App/App.js [hunk 22]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { Routes, Route } from 'react-router-dom';
+import {Provider} from 'react-redux'
+
+import './App.css';
+import Login from '../Pages/Login';
+import SignUp from '../Pages/SignUp';
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/App/App.js]: Please look out lazy loading concepts and add lazy loading too your components.

**Review by RanaWaqar7550 — CHANGES_REQUESTED** [review]

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js (hunk 16 above)
Comment (HassamAzam) [reply Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: I used in the start of the functional component to get an Array. If I use a hook inside a handler function, it gives error that hooks are only called components or inside other custom hook.
Is there another approach to this?

File: Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js [hunk 23]
```diff
@@ -7,12 +8,18 @@ const ButtonComponent = ({
...
+  </Button>
+);
+
 
+Button.propTypes = {
+  label: PropTypes.string.isRequired,  
+   
+};
+Button.defaultProps = {
+  label:"Button Name"
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/ButtonComponent.js]:
```suggestion
label: "Button Name"
```

File: Assignment-10-August-22/userprofile/src/Components/InputField.js [hunk 24]
```diff
@@ -1,7 +1,7 @@
 import React from "react";
 import { TextField } from "@mui/material";
-const InputField = ({ label, placeholder, type = "text", ...props }) => {
-  return (
+
+const InputField = ({ label, placeholder, type = "text", ...props }) => (
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/Components/InputField.js:5]: missing proptypes, as this is a common component.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 25]
```diff
@@ -1,63 +1,84 @@
...
   const navigate = useNavigate();
 
-  const userEmail = useSelector((state) => state.auth.userEmail);
-  const userArray = useSelector((state) => state.userArray.value);
-
-  const loggedInUser = userArray.find((user) => user.email === userEmail);
+  const loggedInUser = useSelector(selectLoggedInUser);
+  const userArray = useSelector(selectUserArray);
 
+  console.log("Logged user", loggedInUser);
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: console?

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 26]
```diff
@@ -1,63 +1,84 @@
...
   const handleLogout = () => {
     dispatch(logout());
     navigate("/login");
   };
 
+  const checkExistence = (userArray,email) =>
+     userArray.some((user) => user.email === email);
+
   const handleUpdate = (values) => {
+    console.log("inside  the update functions");
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: ?

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 27]
```diff
@@ -1,63 +1,84 @@
...
     if (loggedInUser) {
+
+      if (values.email!==loggedInUser.email  && checkExistence(userArray,values.email))
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: we need to use linter properly with our code. 
Add to your IDE or run linter command before commit.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js [hunk 28]
```diff
@@ -105,6 +127,15 @@ export default function Dashboard() {
                     value={values.lastName}
                     onChange={handleChange}
                   />
+                  <InputField
+                    label=""
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Dashboard.js]: missing label field

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js (hunk 16 above)
Comment (RanaWaqar7550) [reply Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: create a utility function and returns user data by giving email.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Login.js [hunk 29]
```diff
@@ -1,46 +1,59 @@
...
+  useEffect(() => {
+    setDocumentTitle("Login");
+  }, []);
+
   const handleLogin = (values) => {
     const user = userArray.find((user) => user.email === values.email);
 
     if (user) {
-      if (user.password === values.password) {
+      if (user?.password === values?.password) {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Login.js]: by adding this above user check can be removed.

File: Assignment-10-August-22/userprofile/src/containers/Pages/Title.js [hunk 30]
```diff
@@ -0,0 +1,6 @@
+const setDocumentTitle = (title) => {
+    
+    document.title = title;
+
+}
+export default setDocumentTitle;
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Title.js]: EOF

File: Assignment-10-August-22/userprofile/src/containers/Pages/Title.js [hunk 31]
```diff
@@ -0,0 +1,6 @@
+const setDocumentTitle = (title) => {
+
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Title.js]: Remove Unnecessary lines

File: Assignment-10-August-22/userprofile/src/containers/Pages/Title.js [hunk 32]
```diff
@@ -0,0 +1,6 @@
+const setDocumentTitle = (title) => {
```
Comment (RanaWaqar7550) [inline Assignment-10-August-22/userprofile/src/containers/Pages/Title.js]: this is a utility function. create a utils folder under pages and move this file there and make file name with camelCase letter or small letters if its not a component.

**Review by RanaWaqar7550 — CHANGES_REQUESTED** [review]

**Review by RanaWaqar7550 — APPROVED** [review]

**Pull request merged**